arelle/plugin/validate/DBA/PluginValidationDataExtension.py

```diff
@@ -94,33 +112,32 @@ def contextFactMap(self, modelXbrl: ModelXbrl) -> dict[str, dict[QName, ModelFac
         return self._contextFactMap
 
-    def getCurrentAndPreviousReportingPeriodContexts(self, modelXbrl: ModelXbrl) -> list[ModelContext]:
-        """
-        :return: Returns the most recent reporting period contexts (at most two).
-        """
-        contexts = self.getReportingPeriodContexts(modelXbrl)
-        if not contexts:
-            return contexts
-        if len(contexts) > 2:
-            return contexts[-2:]
-        return contexts
 
-    def getReportingPeriodContexts(self, modelXbrl: ModelXbrl) -> list[ModelContext]:
+    def getReportingPeriodContexts(self, modelXbrl: ModelXbrl) -> dict[str, Any] | None:
         """
         :return: A sorted list of contexts that match "reporting period" criteria.
         """
         if self._reportingPeriodContexts is not None:
             return self._reportingPeriodContexts
-        contexts = []
-        for context in modelXbrl.contexts.values():
-            context = cast(ModelContext, context)
-            if context.isInstantPeriod or context.isForeverPeriod:
-                continue  # Reporting period contexts can't be instant/forever contexts
+        contexts = {}
+        allContexts = modelXbrl.contexts.values()
+        currentPeriodEndFact = getFirstValidFact(modelXbrl, self.reportingPeriodEndDateQn)
+        currentPeriodStartFact = getFirstValidFact(modelXbrl, self.reportingPeriodStartDateQn)
+        pastPeriodEndFact = getFirstValidFact(modelXbrl, self.precedingReportingPeriodEndDateQn)
+        pastPeriodStartFact = getFirstValidFact(modelXbrl, self.precedingReportingPeriodStartDateQn)
+        for context in allContexts:
             if len(context.qnameDims) > 0:
                 if context.qnameDims.keys() != {self.consolidatedSoloDimensionQn}:
                     continue  # Context is dimensionalized with something other than consolidatedSoloDimensionQn
                 if context.dimMemberQname(self.consolidatedSoloDimensionQn) != self.consolidatedMemberQn:
                     continue  # Context is dimensionalized with the correct dimension but not member
-            contexts.append(context)
-        self._reportingPeriodContexts = sorted(contexts, key=lambda c: c.endDatetime)
+            contextEndDate = context.endDatetime - timedelta(days=1)  # The endDatetime on the context is a day later than the actual endDatetime as defined in the XBRL.
+            if currentPeriodEndFact is not None and currentPeriodStartFact is not None:
+                if context.startDatetime == currentPeriodStartFact.xValue and contextEndDate == currentPeriodEndFact.xValue:
+                    contexts["current"] = context
+                    continue
+            if pastPeriodEndFact is not None and  pastPeriodStartFact is not None:
+                if context.startDatetime  == pastPeriodStartFact.xValue and contextEndDate == pastPeriodEndFact.xValue:
+                    contexts["previous"] = context
```

Comment:
The magic `current` and `previous` dict keys would be better replaced with typed fields of a small dataclass or named tuple so that callers of this function don't have to inspect the implementation to understand the return type:

```python
@dataclass(frozen=True)
class ReportingPeriodContexts:
    current: set[ModelContext]
    previous: set[ModelContext]
```